### PR TITLE
WIP: Parquet Reader Optimizations

### PR DESCRIPTION
**Pull request by vkorukanti**
Adds new ColumnReaders for types: INT32, INT64, INT96, VarBinary, Boolean (nested, array, and non-nested). Remaining: Fixed Len Binary, Decimal
Supported encodings: RLE, PACKED, DICTIONARY. Delta encodings support is not yet added.

Main changes in the new column readers are:
 * Batch reads from Decoders
 * Use ArrayBlock implementations instead of BlockBuilder
 * New RLE Decoders that contains both the dictionary and values decoder. Dictionary lookup only once for the RLE block.
 * Separate non-nested and nested column readers
    * non-Nested columns don’t need RL
    * non-Nested and non-nullable columns don’t need to read DL
 * Avoid using temp byte arrays in case of VarBinary
 * Skip values without decoding

Also adds the benchmarks (based on ORC benchmarking code)

Benchmark | Mode | Old | Error | New | Error
-- | -- | -- | -- | -- | --
readBooleanNoNull | avgt | 39.44 | ±0.702 | 1.523 | ±0.046
readBooleanWithNull | avgt | 52.746 | ±1.459 | 11.511 | ±0.239
readInt32NoNull | avgt | 59.116 | ±1.423 | 9.968 | ±1.031
readInt32WithNull | avgt | 60.933 | ±1.277 | 16.179 | ±0.32
readInt64NoNull | avgt | 77.31 | ±3.451 | 20.309 | ±1.277
readInt64WithNull | avgt | 67.296 | ±1.5 | 21.531 | ±0.548
readInt96NoNull | avgt | 126.638 | ±5.213 | 44.831 | ±2.071
readInt96WithNull | avgt | 93.29 | ±2.707 | 33.348 | ±0.767
readListBooleanWithNull | avgt | 159.107 | ±2.767 | 96.307 | ±1.79
readListInt32WithNull | avgt | 171.035 | ±3.819 | 99.612 | ±1.651
readListInt64WithNull | avgt | 185.385 | ±4.309 | 109.098 | ±3.58
readListInt96WithNull | avgt | 227.756 | ±5.58 | 127.24 | ±3.924
readListSliceDictionaryWithNull | avgt | 188.426 | ±3.223 | 105.013 | ±2.163
readSliceDictionaryNoNull | avgt | 76.132 | ±1.778 | 15.625 | ±0.324
readSliceDictionaryWithNull | avgt | 72.929 | ±1.871 | 18.103 | ±0.358
readStructBooleanWithNull | avgt | 61.284 | ±1.271 | 28.969 | ±0.784
readStructInt32WithNull | avgt | 71.697 | ±1.827 | 35.646 | ±0.762
readStructInt64WithNull | avgt | 78.999 | ±1.826 | 42.849 | ±0.769
readStructInt96WithNull | avgt | 104.609 | ±2.012 | 54.361 | ±1.225
readStructSliceDictionaryWithNull | avgt | 85.545 | ±1.879 | 38.806 | ±0.92

```
== RELEASE NOTES ==

General Changes
* ...
* ...

Hive Changes
* ...
* ...
```

If release note is NOT required, use:

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/Int32SimpleColumnReaderV2.java [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+        while (remainingInBatch > 0) {
+            if (remainingCountInPage == 0) {
+                if (!readNextPage()) {
+                    break;
+                }
+            }
+
+            int readChunkSize = Math.min(remainingCountInPage, remainingInBatch);
+
+            Int32ValuesDecoder valuesDecoder = (Int32ValuesDecoder) this.valuesDecoder;
```
Comment (zhenxiao) [inline presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/Int32SimpleColumnReaderV2.java]: Presto generally not use this to reference data member

File: presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/Int32SimpleColumnReaderV2.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+            startOffset += readChunkSize;
+            remainingInBatch -= readChunkSize;
+            remainingCountInPage -= readChunkSize;
+        }
+
+        if (remainingInBatch != 0) {
+            throw new ParquetDecodingException("Still remaining to be read in current batch.");
+        }
+
+        Block block = new IntArrayBlock(nextBatchSize, Optional.empty(), values);
```
Comment (zhenxiao) [inline presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/Int32SimpleColumnReaderV2.java]: static import Optional.empty()

File: presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/decoders/Decoders.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+                case FIXED_LEN_BYTE_ARRAY:
+                default:
+                    throw new PrestoException(PARQUET_UNSUPPORTED_COLUMN_TYPE, format("Column: %s, Encoding: %s", columnDescriptor, encoding));
+            }
+        }
+
+        throw new PrestoException(PARQUET_UNSUPPORTED_ENCODING, format("Column: %s, Encoding: %s", columnDescriptor, encoding));
+    }
+
+    public static final SimpleDLDecoder createSimpleDLDecoder(ParquetEncoding encoding, boolean isRequiredType, int maxLevelValue, int valueCount, ByteBuffer buffer)
```
Comment (zhenxiao) [inline presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/decoders/Decoders.java]: let's try remove simple everywhere:
s/SimpleDLDecoder/DefinitionLevelDecoder/g
s/createSimpleDLDecoder/createDefinitionLevelDecoder/g

File: presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/decoders/rle/BaseRLEBitPackedDecoder.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+import static com.facebook.presto.parquet.readerv2.decoders.rle.BaseRLEBitPackedDecoder.Mode.PACKED;
+import static com.facebook.presto.parquet.readerv2.decoders.rle.BaseRLEBitPackedDecoder.Mode.RLE;
+
+public abstract class BaseRLEBitPackedDecoder
+{
+    private final boolean rleOnlyMode;
+
+    private final int bitWidth;
+    private final BytePacker packer;
+    private final InputStream in;
```
Comment (zhenxiao) [inline presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/decoders/rle/BaseRLEBitPackedDecoder.java]: s/in/inputStream/g

File: presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/decoders/rle/BaseRLEBitPackedDecoder.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+            // for RLE only mode there is nothing more to read
+            return false;
+        }
+
+        if (in.available() <= 0) {
+            this.currentCount = 0;
+            return false;
+        }
+
+        int header = BytesUtils.readUnsignedVarInt(this.in);
```
Comment (zhenxiao) [inline presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/decoders/rle/BaseRLEBitPackedDecoder.java]: static import readUnsignedVarInt
remove this

File: presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/decoders/rle/BaseRLEBitPackedDecoder.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        this.mode = (header & 1) == 0 ? RLE : PACKED;
+        switch (this.mode) {
+            case RLE:
+                this.currentCount = header >>> 1;
+                this.currentValue = BytesUtils.readIntLittleEndianPaddedOnBitWidth(this.in, this.bitWidth);
```
Comment (zhenxiao) [inline presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/decoders/rle/BaseRLEBitPackedDecoder.java]: static import BytesUtils.readIntLittleEndianPaddedOnBitWidth
do not use this to reference data members

File: presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/decoders/rle/BaseRLEBitPackedDecoder.java [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+                return true;
+            case PACKED:
+                int numGroups = header >>> 1;
+                this.currentCount = numGroups * 8;
+                this.currentBuffer = new int[this.currentCount];
+                byte[] bytes = new byte[numGroups * this.bitWidth];
+                int bytesToRead = (int) Math.ceil((double) (this.currentCount * this.bitWidth) / 8.0D);
```
Comment (zhenxiao) [inline presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/decoders/rle/BaseRLEBitPackedDecoder.java]: static import Math.ceil

File: presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/decoders/rle/BaseRLEBitPackedDecoder.java [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+                bytesToRead = Math.min(bytesToRead, this.in.available());
+                (new DataInputStream(this.in)).readFully(bytes, 0, bytesToRead);
```
Comment (zhenxiao) [inline presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/decoders/rle/BaseRLEBitPackedDecoder.java]: have a variable to hold new DataInputStream

File: presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/decoders/SimpleDLDecoder.java [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+                    break;
+                }
+                default:
+                    throw new ParquetDecodingException("not a valid mode " + this.mode);
+            }
+            this.currentCount -= readChunkSize;
+            remainingToCopy -= readChunkSize;
+        }
+
+        checkState(remainingToCopy == 0, "Failed to copy the requested number of DLs");
```
Comment (zhenxiao) [inline presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/decoders/SimpleDLDecoder.java]: DLs -> definition levels

File: presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/decoders/SimpleDLDecoder.java [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    public SimpleDLDecoder(int rleValue, int valueCount)
+    {
+        super(rleValue, valueCount);
+    }
+
+    public int readNext(boolean[] values, int offset, int length)
+            throws IOException
+    {
+        int nonNullCount = 0;
+        int destIndex = offset;
```
Comment (zhenxiao) [inline presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/decoders/SimpleDLDecoder.java]: destIndex -> destinationIndex or destinationPosition
srcIdx -> sourcePosition
remove this

File: presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/decoders/plain/Int32PlainValuesDecoder.java [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+    {
+        checkArgument(bufferOffset + length * 4 <= bufferEnd, "End of stream: invalid read request");
+        checkArgument(length >= 0 && offset >= 0, "invalid read request: offset %s, length", offset, length);
+
+        final int endOffset = offset + length;
+        final byte[] localByteBuffer = byteBuffer;
+        int localBufferOffset = bufferOffset;
+
+        while (offset < endOffset) {
+            values[offset++] = BytesUtils.getInt(localByteBuffer, localBufferOffset);
```
Comment (zhenxiao) [inline presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/decoders/plain/Int32PlainValuesDecoder.java]: static import BytesUtils.getInt

File: presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/AbstractNestedColumnReaderV2.java [hunk 12]
```diff
@@ -0,0 +1,282 @@
...
+import java.util.List;
+
+import static com.facebook.presto.parquet.readerv2.decoders.Decoders.createDLDecoder;
+import static com.facebook.presto.parquet.readerv2.decoders.Decoders.createRLDecoder;
+import static com.facebook.presto.parquet.readerv2.decoders.Decoders.createValuesDecoder;
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+import static org.apache.parquet.bytes.BytesUtils.getWidthFromMaxInt;
+
+public abstract class AbstractNestedColumnReaderV2
```
Comment (zhenxiao) [inline presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/AbstractNestedColumnReaderV2.java]: s/AbstractNestedColumnReaderV2/AbstractNestedColumnReader/g

File: presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/BinaryNestedColumnReaderV2.java [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+            ReadChunk readChunk = readChunkList.get(i);
+
+            bufferIdx = ((BinaryValuesDecoder) valuesDecoderInfo.getValuesDecoder()).readIntoBuffer(byteBuffer, bufferIdx, offsets, offsetIdx, readChunk);
+            offsetIdx += valuesDecoderInfo.getValueCount();
+            i++;
+        }
+
+        boolean[] isNull = new boolean[newBatchSize];
+
+        //Collections.reverse(dlDecodingInfo.getValuesDecoderInfos());
```
Comment (zhenxiao) [inline presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/BinaryNestedColumnReaderV2.java]: remove?

**Comment by zhenxiao** [comment]
BTW, could you please resolve code conflicts?

File: presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/AbstractSimpleColumnReaderV2.java [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+        else {
+            this.dlDecoder = new SimpleDLDecoder(valueCount, new ByteArrayInputStream(pageV2.getDefinitionLevels().getBytes()));
+        }
+
+        final byte[] dataBuffer = pageV2.getSlice().getBytes();
+        this.valuesDecoder = createValuesDecoder(columnDescriptor, dictionary, pageV2.getDataEncoding(), dataBuffer, 0, dataBuffer.length);
+    }
+
+    protected abstract ColumnChunk readNoNull()
+            throws IOException;
```
Comment (zhenxiao) [inline presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/AbstractSimpleColumnReaderV2.java]: please move all abstract methods declaration above implementations, just below constructor implementation

File: presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/AbstractSimpleColumnReaderV2.java [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+        return true;
+    }
+
+    private void readPageV1(DataPageV1 page)
+            throws IOException
+    {
+        byte[] bytes = page.getSlice().getBytes();
+        ByteBuffer byteBuffer = ByteBuffer.wrap(bytes, 0, bytes.length);
+
+        this.dlDecoder = createSimpleDLDecoder(page.getDefinitionLevelEncoding(), columnDescriptor.isRequired(), columnDescriptor.getMaxDefinitionLevel(), page.getValueCount(), byteBuffer);
```
Comment (zhenxiao) [inline presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/AbstractSimpleColumnReaderV2.java]: remove this when referencing data members
s/dlDecoder/definitionLevelDecoder/g

File: presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/AbstractSimpleColumnReaderV2.java [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+        this.valuesDecoder = createValuesDecoder(columnDescriptor, dictionary, page.getValueEncoding(),
+                bytes, byteBuffer.position(), bytes.length - byteBuffer.position());
+    }
+
+    private void readPageV2(DataPageV2 pageV2)
+            throws IOException
+    {
+        final int valueCount = pageV2.getValueCount();
+
+        final int maxDL = columnDescriptor.getMaxDefinitionLevel();
```
Comment (zhenxiao) [inline presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/AbstractSimpleColumnReaderV2.java]: s/maxDL/maxDefinitionLevel/g

File: presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/Int32SimpleColumnReaderV2.java [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+        return new ColumnChunk(block, new int[0], new int[0]);
+    }
+
+    @Override
+    protected void skip(int skipSize)
+            throws IOException
+    {
```
Comment (zhenxiao) [inline presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/Int32SimpleColumnReaderV2.java]: int remainingInBatch = skipSize;
and remove the inner definitions in both if cases

File: presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/decoders/Decoders.java [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+import static org.apache.parquet.bytes.BytesUtils.getWidthFromMaxInt;
+import static org.apache.parquet.bytes.BytesUtils.readIntLittleEndianOnOneByte;
+
+public class Decoders
+{
+    private Decoders()
+    {
+    }
+
+    public static final ValuesDecoder createValuesDecoder(ColumnDescriptor columnDescriptor, Dictionary dictionary, ParquetEncoding encoding,
```
Comment (zhenxiao) [inline presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/decoders/Decoders.java]: all parameters in same line, or
each in one line

File: presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/decoders/rle/Int32RLEDictionaryValuesDecoder.java [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    @Override
+    public void readNext(int[] values, int offset, int length)
+            throws IOException
+    {
+        int destIndex = offset;
+        int remainingToCopy = length;
+        while (remainingToCopy > 0) {
+            if (this.currentCount == 0) {
```
Comment (zhenxiao) [inline presto-parquet/src/main/java/com/facebook/presto/parquet/readerv2/decoders/rle/Int32RLEDictionaryValuesDecoder.java]: access data member without this, same for all files

File: presto-parquet/src/test/java/com/facebook/presto/parquet/readerv2/TestColumnReaders.java [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+
+import com.facebook.presto.parquet.BenchmarkParquetReader;
+import org.testng.annotations.Test;
+
+public class TestColumnReaders
+{
+    @Test
+    public void columnReaders()
+    {
+        // Test end-2-end column reader using the BenchmarkParquetReader, it already contains
```
Comment (zhenxiao) [inline presto-parquet/src/test/java/com/facebook/presto/parquet/readerv2/TestColumnReaders.java]: remove this comment?

File: presto-parquet/src/main/java/com/facebook/presto/parquet/reader/ParquetReader.java [hunk 21]
```diff
@@ -249,6 +257,7 @@ private ColumnChunk readPrimitive(PrimitiveField field)
         }
 
         ColumnChunk columnChunk = columnReader.readNext();
+        columnChunk = rewriteIfNeeded(columnChunk, field.getDescriptor().getPrimitiveType().getPrimitiveTypeName(), field.getType());
```
Comment (zhenxiao) [inline presto-parquet/src/main/java/com/facebook/presto/parquet/reader/ParquetReader.java:260]: s/rewriteIfNeeded/typeCoercion/g

**Comment by zhenxiao** [comment]
is this related:
[ERROR] Tests run: 148, Failures: 1, Errors: 0, Skipped: 147, Time elapsed: 37.017 s <<< FAILURE! - in TestSuite
[ERROR] init(com.facebook.presto.hive.TestHivePushdownIntegrationSmokeTest)  Time elapsed: 36.416 s  <<< FAILURE!
java.lang.RuntimeException: 2
	at com.facebook.presto.tests.AbstractTestingPrestoClient.execute(AbstractTestingPrestoClient.java:115)
	at com.facebook.presto.tests.DistributedQueryRunner.execute(DistributedQueryRunner.java:436)
	at com.facebook.presto.tests.QueryAssertions.copyTable(QueryAssertions.java:364)
	at com.facebook.presto.tests.QueryAssertions.copyTable(QueryAssertions.java:356)
	at com.facebook.presto.tests.QueryAssertions.copyTpchTables(QueryAssertions.java:348)
	at com.facebook.presto.tests.QueryAssertions.copyTpchTables(QueryAssertions.java:334)
	at com.facebook.presto.hive.HiveQueryRunner.createQueryRunner(HiveQueryRunner.java:180)
	at com.facebook.presto.hive.HiveQueryRunner.createQueryRunner(HiveQueryRunner.java:110)
	at com.facebook.presto.hive.TestHivePushdownIntegrationSmokeTest.lambda$new$0(TestHivePushdownIntegrationSmokeTest.java:37)
	at com.facebook.presto.tests.AbstractTestQueryFramework.init(AbstractTestQueryFramework.java:85)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.Invoker.invokeConfigurationMethod(Invoker.java:515)
	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:217)
	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:144)
	at org.testng.internal.TestMethodWorker.invokeBeforeClassMethods(TestMethodWorker.java:169)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ArrayIndexOutOfBoundsException: 2
	at com.google.common.collect.RegularImmutableList.get(RegularImmutableList.java:60)
	at com.facebook.presto.execution.scheduler.FixedBucketNodeMap.getAssignedNode(FixedBucketNodeMap.java:44)
	at com.facebook.presto.execution.scheduler.BucketNodeMap.getAssignedNode(BucketNodeMap.java:48)
	at com.facebook.presto.execution.scheduler.NodeScheduler.selectDistributionNodes(NodeScheduler.java:303)
	at com.facebook.presto.execution.scheduler.nodeSelection.SimpleNodeSelector.computeAssignments(SimpleNodeSelector.java:199)
	at com.facebook.presto.execution.scheduler.FixedSourcePartitionedScheduler$BucketedSplitPlacementPolicy.computeAssignments(FixedSourcePartitionedScheduler.java:316)
	at com.facebook.presto.execution.scheduler.SourcePartitionedScheduler.schedule(SourcePartitionedScheduler.java:273)
	at com.facebook.presto.execution.scheduler.FixedSourcePartitionedScheduler.schedule(FixedSourcePartitionedScheduler.java:240)
	at com.facebook.presto.execution.scheduler.LegacySqlQueryScheduler.schedule(LegacySqlQueryScheduler.java:348)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	... 3 more

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   TestHivePushdownIntegrationSmokeTest>AbstractTestQueryFramework.init:85->lambda$new$0:37 Â» Runtime
[INFO] 
[ERROR] Tests run: 148, Failures: 1, Errors: 0, Skipped: 147
[INFO]

File: presto-parquet/src/main/java/com/facebook/presto/parquet/reader/AbstractColumnReader.java [hunk 22]
```diff
@@ -26,7 +26,12 @@
 import com.facebook.presto.parquet.ParquetTypeUtils;
 import com.facebook.presto.parquet.RichColumnDescriptor;
 import com.facebook.presto.parquet.dictionary.Dictionary;
+<<<<<<< HEAD
```
[the remark this answers is not in the document]
Comment (zhenxiao) [reply presto-parquet/src/main/java/com/facebook/presto/parquet/reader/AbstractColumnReader.java]: Hi @vkorukanti currently we will merge each commit in a PR as separate commits into master. Would be nice if we could keep each commit clean. Could you please get this resolved in this commit?

**Comment by zhenxiao** [comment]
hi @vkorukanti currently we will merge individual commits in a PR as separate commits into master. Would be nice if we could keep each commit clean. Could you please resolve file merge in: https://github.com/prestodb/presto/pull/13638/commits/ddc8a6688a53cc1669fe48feb75c1775ce332e9f

**Comment by vkorukanti** [comment]
> hi @vkorukanti currently we will merge individual commits in a PR as separate commits into master. Would be nice if we could keep each commit clean. Could you please resolve file merge in: [ddc8a66](https://github.com/prestodb/presto/commit/ddc8a6688a53cc1669fe48feb75c1775ce332e9f)
Just force pushed, this should be fixed. The failed test ran fine in locally `mvn test -Ptest-hive-pushdown-filter-queries-basic`. Also it passed in latest CI run.

**Pull request merged**